yetanotherco/aligned_layer
# feat: integrate batcher with payment system

Integrates batcher with the new batcher payment service.
When a user sends a proof, the message is signed and the batcher verifies the signature and recovers the user address. Then, it checks that the address has funds in the batcher payment service.

### Testing:
 For fine grained testing:
 * Start anvil with block time, aggregator and operator
 * Start batcher with
 ```bash 
cargo +nightly-2024-04-17 run --manifest-path ./batcher/aligned-batcher/Cargo.toml --release -- --config ./config-files/config-batcher.yaml --env-file ./batcher/aligned-batcher/.env
```
This will start the batcher without sending funds.
* Send some proofs with
```bash
aligned submit --proving_system GnarkPlonkBls12_381 \
--proof batcher/aligned/test_files/plonk_bls12_381/plonk.proof \
--public_input batcher/aligned/test_files/plonk_bls12_381/plonk_pub_input.pub \
--vk batcher/aligned/test_files/plonk_bls12_381/plonk.vk --repetitions 2 --keystore_path config-files/anvil.ecdsa.key.json
```
An error will be logged meaning that the user has no funds in the batcher payment system.
* Send some funds with
```bash
cast send 0x7969c5eD335650692Bc04293B07F5BF2e7A673C0 --value 1ether --rpc-url http://localhost:8545 --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80
```
* Send proofs again, everything should work as usual.

## Happy path:
* Start anvil again so that the previous state is reseted, and do the usual flow. Everything should work as usual. The only difference now is that when `make batcher_start` is called, a transaction is made from the user to the payment system to add funds in devnet.
